### PR TITLE
Have fuzzy matching on user supplied parameters and paths

### LINKED ISSUE #373: Have fuzzy matching on user supplied parameters and paths
In the `prepare_run.py` when a user supplied parameters is not correct it would be very nice if the parameter is matched against all possible parameters and the closest one is given back to the user as a suggestion.

It could look like the following example:
```
Error 01: Receptor file not found

    |  ~/project_01/files/receptor.pdb

help= the folder 'project_01' does not live in '~' did you mean 'project-01'?
```

Discussed with @joaomcteixeira.

### PR DESCRIPTION
**Pull request by douweschulte**
Closes #373.

You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and you comply with the following criteria:

- [x] You have stick to Python. Talk with us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [x] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there's a (state) purpose
- [x] code follows our coding style
- [x] You wrote tests for the new code
- [ ] `tox` tests pass. *Run `tox` command inside the repository folder*
- [ ] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [x] PR does not add any *install dependencies* unless permission granted by the HADDOCK team
- [x] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

As discussed yesterday. This PR adds nice error messages for users to see when a parameter/file is mistyped what parameter/file/path is the closest to their supplied parameter/file. This enhances the user experience because it is much easier to spot the difference between `project_01` and `project_o1` if it is printed out on your screen.

### CODE REVIEW

**Comment by douweschulte** [comment]
With all these nice lints is it not also time for a spell checker lint? 
image

File: src/haddock/gear/prepare_run.py [hunk 1]
```diff
@@ -219,9 +220,16 @@ def validate_modules_params(modules_params):
             - expandable_params
 
         if diff:
+            matched = fuzzy_match(diff, defaults.keys())
+
+            def pretty_print(match):
+                return f" * \'{match[0]}\' did you mean \'{match[1]}\'?"
+
+            eol = '\n'
```
Comment (joaomcteixeira) [inline src/haddock/gear/prepare_run.py:228]: in python:
```python
import os
os.linesep.join(...)
```
could you change? This get's you OS-specific line break character.

File: src/haddock/gear/prepare_run.py [hunk 2]
```diff
@@ -517,3 +530,113 @@ def populate_mol_parameters(modules_params):
...
+    ------
+    ValueError
+        If the path does not exist.
+    """
+    path = os.path.normpath(path)
+    if os.path.exists(path):
+        return None
+
+    split = os.path.split(path)
+    elements = split[0].split(os.sep) + split[1]
```
Comment (joaomcteixeira) [inline src/haddock/gear/prepare_run.py:562]: Do you mind refactoring to using `pathlib` ? The community is moving towards using it always when handling paths.
```python
elements = Path(path).parts
for part in elements:
    next_folder = Path(reconstitutied_path, part)
    if not next_folder.exists():
         ....

File: src/haddock/gear/prepare_run.py [hunk 3]
```diff
@@ -517,3 +530,113 @@ def populate_mol_parameters(modules_params):
...
+    Returns
+    -------
+    list(string, string)
+        The closest string from the possibilities to each string of the
+        `user_input`. With as first element of the tuple the user_input
+        string, and as second element the matched possibility.
+    """
+    results = list()
+
+    for user_word in user_input:
```
Comment (joaomcteixeira) [inline src/haddock/gear/prepare_run.py:600]: You use our nice function:

```
from haddock.libs.libutil import transform_to_list
....
for user in transform_to_list(user_input):
...
```
like that you can call the function directly with a string.

File: src/haddock/gear/prepare_run.py [hunk 4]
```diff
@@ -517,3 +530,113 @@ def populate_mol_parameters(modules_params):
...
+            break
+
+    msg = (f"The following file could not be found: \'{path}\'."
+           f"In the folder \'{error[0]}\' the following \'{error[1]}\' "
+           f"is the closest match to the supplied \'{error[2]}\', did "
+           "you mean to open this?")
+    raise ValueError(msg)
+
+
+def fuzzy_match(user_input, possibilities):
```
Comment (joaomcteixeira) [inline src/haddock/gear/prepare_run.py:579]: move this function to `libs.libmath` please.

File: src/haddock/gear/prepare_run.py [hunk 5]
```diff
@@ -517,3 +530,113 @@ def populate_mol_parameters(modules_params):
...
+        for possibility in possibilities:
+            distance = levenshtein_distance(user_input, possibility)
+            if distance < best[0]:
+                best = (distance, possibility)
+        results += (user_word, best[1])
+
+    return results
+
+
+def levenshtein_distance(left, right):
```
Comment (joaomcteixeira) [inline src/haddock/gear/prepare_run.py:611]: move this function to `libs.libmath`.

**Review by joaomcteixeira — CHANGES_REQUESTED** [review]
Very nice!! @douweschulte 
small details.

**Comment by joaomcteixeira** [comment]
Closing because we saw using `difflib` is much faster. See #381

**Pull request closed without merging**